Fix several correctness issues in TempRValueOpt

eeckstein opened the pull request:
* fix the handling of begin_access
* fix the handling of load [take]
* don't allow copy_addr [take] from a projection of the temporary.
* fix the placement of destroy_addr in case of a copy_addr [take]

plus some refactoring.
For details see the commit messages.

rdar://problem/69757314

eeckstein committed:
MemBehavior: handle begin_access when checking for apply side-effects.

eeckstein committed:
TempRValueOpt: don't use collectLoads in tryOptimizeStoreIntoTemp

This refactoring removes a lot of special-casing in collectLoads and also makes tryOptimizeStoreIntoTemp simpler.
It's a NFC.

eeckstein committed:
TempRValueOpt: refactoring: simplify the signature of collectLoads

It's sufficient to pass the operand instead of the operand, the user and the operand value.
NFC.

eeckstein committed:
TempRValueOpt: move the mayWrite-check for applies from collectLoads to checkNoSourceModification

... where it belongs.
This is mostly refactoring, but it also fixes a bug: we don't recurse into a begin_access in collectLoads.
If there is an apply in such a scope, the mayWrite-check wouldn't be done.
In checkNoSourceModification all instructions are visited, so the check is always done.

eeckstein committed:
TempRValueOpt: fix the handling of begin_access

Consider the related end_access instructions as uses to correctly mark the end of the lifetime of the temporary.
This fixes a miscompile in case there is a modification of the copy-source between an begin_access and end_access.

eeckstein committed:
TempRValueOpt: fix the handling of load [take]

load [take] was not considered as a use and it was not detected if it's in a different basic block.
This fixes a miscompile in case there is a modification of the copy-source before a load [take].

rdar://problem/69757314

eeckstein committed:
TempRValueOpt: fix the placement of destroy_addr in case of a copy_addr [take]

Instead of reusing the existing destroy_addr (or load/copy_addr [take]) of the temporary, insert a new destroy_addr - at the correct location.
This fixes a memory lifetime failure in case the copy-source is modified (actually: re-initialized) after the last use of the temporary: E.g. (before copy elimination):

copy_addr [take] %src to [initialization] %temp
%x = load %temp // last use of %temp
store %y to [init] %src
destroy_addr %temp

The correct location for the destroy_addr is after the last use (after copy elimination):

%x = load %src
destroy_addr %src
store %y to [init] %src

rdar://problem/69757314

eeckstein committed:
TempRValueOpt: don't allow copy_addr [take] from a projection of the temporary.

This fixes a memory lifetime failure.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

gottesmm commented on test/SILOptimizer/temp_rvalue_opt_ossa.sil:
```diff
@@ -180,6 +185,26 @@ bb0(%0 : $*Builtin.NativeObject):
+  %a0 = struct_element_addr %0 : $*Two, #Two.a
+  %b0 = struct_element_addr %0 : $*Two, #Two.b
+  %s = alloc_stack $Two
+  copy_addr [take] %1 to [initialization] %s : $*Two
```
Is this an issue in non-ossa code as well?

eeckstein replied:
yes. The only difference is that the memory lifetime verifier will not catch it (because it only runs on ossa).

eeckstein replied:
The non-ossa version would be completely equivalent and also be handled equivalently in the optimization.

gottesmm replied:
Ok.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test linux

gottesmm commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -441,7 +467,18 @@ bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
+  if (!lastLoadInst)
+    return false;
+
+  bool needToInsertDestroy = copyInst->isTakeOfSrc();
```
I would also add that in a comment that we are going to insert a destroy_addr in the appropriate spot after rewriting, but before deleting instructions rather than rewrite instructions to ensure proper destruction.

eeckstein replied:
I added a comment

gottesmm commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -478,43 +508,17 @@ bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
       if (cai != copyInst) {
         assert(cai->getSrc() == tempObj);
-        if (cai->isTakeOfSrc() && !copyInst->isTakeOfSrc())
+        if (cai->isTakeOfSrc())
```
I would add a comment here saying:

"If we have a copy addr user that is not our original copy inst, since our copy_addr passes in a +1 value, if we have a take, convert it to be a non-take. If we do a take of the source of our original copy src, we will handle it by inserting a destroy_addr in the appropriate place after this loop."

eeckstein replied:
I explained it in the comment above the loop

gottesmm commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -291,9 +291,15 @@ collectLoads(Operand *addressUse, CopyAddrInst *originalCopy,
+/// Return the instrution which is the last use (or copyInst if there are no
+/// uses).
+/// Returns a nullptr, if there is a source modification before the last use.
+SILInstruction *TempRValueOptPass::checkNoSourceModification(
```
Can we rename this now/make the comment clearer. My thought is the comment should be something like:

```
Given the list of uses of the target copy_addr that was found by
walking def-use chains, walk the block where those uses were
defined to prove that the source is not written to before the
last use of that temporary.

Returns the last load in the use block that we found or the
underlying copy_addr we are trying to eliminate if we do not have
any uses. Returns nullptr, if there are any source modifications
before we find the last use in our copy_addr use list.

NOTE: Unfortunately, we cannot simply use the destroy points of
the temporary as the lifetime end, because these destroys may be
in a different basic block due to SILGen cleaning up temporaries
at end of lexical scopes rather than at last use. To work around
this, our caller guarantees that all normal uses are within the
block of the temporary so that we can just look for the last
non-destroying use, which we can use to model the end of the
lifetime of the temporary.
```

I would also change the name of the function to be mean something that implies something more along the lines of the comment. My thought is that the function name should have something that suggests we are performing a walk of the temporaries uses, not just checking no source modification.

eeckstein replied:
I renamed it and change the comment (though I kept it smaller than you suggested)

gottesmm reviewed (commented):
Some initial comments before Erik pushes a fix.

eeckstein committed:
SILBuilder: add an API to insert _after_ an instruction.

... and use that API in FullApplySite::insertAfterInvocation.

Also change FullApplySite::insertAfterInvocation/insertAfterFullEvaluation to directly pass a SILBuilder instead of just an insertion point to the callback.
This makes more sense (given the function names) and simplifies the usages.

It's a NFC.

eeckstein committed:
TempRValueOpt: only do the check for "unusual" temp object destruction in non-OSSA mode.

In OSSA, memory locations are always destroyed in a "visible" way, e.g. with destroy_addr or load [take].

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

eeckstein commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -189,9 +153,24 @@ bool TempRValueOptPass::collectLoads(
+    // result, because a SILAccessKind::Read already guarantees that there are
+    // no writes to the beginAccess result address (or any projection from it).
+    // But we have to register the end-accesses as loads to correctly mark the
+    // end-of-lifetime of the temporary object.
```
done

eeckstein commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -402,6 +408,26 @@ bool TempRValueOptPass::checkTempObjectDestroy(
+/// If the \p lastLoadInst is a terminator (e.g. a try_apply), then destroy_addr
+/// instructions are inserted at the begin of all successor blocks.
+static void insertDestroy(SILValue copySrc, SILInstruction *lastLoadInst) {
+  if (isa<TermInst>(lastLoadInst)) {
```
Good idea! While I added it, i refactored the ApplySite API a bit: 9a10ec7d581345cbfcd15990c54848ca1b25dc8c

eeckstein commented on test/SILOptimizer/temp_rvalue_opt.sil:
```diff
@@ -456,6 +456,34 @@ bb3:
+// CHECK:   store %1 to [init] %0
+// CHECK: } // end sil function 'copy_take_and_try_apply'
+sil [ossa] @copy_take_and_try_apply : $@convention(thin) (@inout Klass, @owned Klass) -> () {
+bb0(%0 : $*Klass, %1 : @owned $Klass):
```
I moved it

eeckstein commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -291,9 +291,15 @@ collectLoads(Operand *addressUse, CopyAddrInst *originalCopy,
 /// lifetime.
-bool TempRValueOptPass::checkNoSourceModification(
+///
+/// Return the instrution which is the last use (or copyInst if there are no
```
done

eeckstein commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -441,7 +467,18 @@ bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
+  
+  // We cannot insert a destroy after lastLoadInst if the copySrc is written
+  // by this instruction. This is a corner case, but can happen if lastLoadInst
+  // is a copy_addr.
```
done

eeckstein commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -441,7 +467,18 @@ bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
   }
 
   // Check if the source is modified within the lifetime of the temporary.
```
It should be clearer now with the new function name

eeckstein commented on lib/SILOptimizer/Transforms/TempRValueElimination.cpp:
```diff
@@ -241,7 +241,8 @@ collectLoads(Operand *addressUse, CopyAddrInst *originalCopy,
                                       originalCopy, loadInsts);
-  case SILInstructionKind::LoadInst:
+
+  case SILInstructionKind::LoadInst: {
```
yes, because we are inserting a new delete_addr now.
